File: tc-v2-verify-go/sign_data.go
// Package tcv2verify is the Go reference for the two TC v2 owner-signature contracts.
//
// Contract A — TC_V2_SIGNDATA_VERIFY_V1 — lives in this file.
// Contract B — TC_V2_TONPROOF_VERIFY_V1 — lives in ton_proof.go.
//
// Implemented from the normative draft (docs/draft/tc-v2-sig-verify-v1-draft.md), NOT
// ported from the TS or Rust implementations — so agreement across the three is evidence
// the draft is unambiguous, not evidence that one port copied another.
//
// The two contracts share NO serialization/endian/hash-pipeline helper and there is no
// universal verifier facade (docs/spec/tc-v2-contract-boundaries.md). Each file owns its
// own encoders. crypto/sha256 and crypto/ed25519 are contract-agnostic primitives.
package tcv2verify

import (
	"crypto/sha256"
	"encoding/base64"
	"encoding/binary"
)

const signDataSchemaPrefix = "ton-connect/sign-data/"

// SignDataInput is the input to Contract A.
type SignDataInput struct {
	Workchain   int32
	AddressHash [32]byte
	Domain      string
	Timestamp   uint64
	PayloadType string // "text" | "binary"
	PayloadText string // used when PayloadType == "text"
	PayloadB64  string // used when PayloadType == "binary" (standard base64)
}

// SignDataDigest builds the Contract A message and returns its sha256 digest (single
// envelope, big-endian length/timestamp, "txt"/"bin" discriminator).
func SignDataDigest(in SignDataInput) [32]byte {
	var tag, payload []byte
	if in.PayloadType == "text" {
		tag = []byte("txt")
		payload = []byte(in.PayloadText)
	} else {
		tag = []byte("bin")
		payload, _ = base64.StdEncoding.DecodeString(in.PayloadB64)
	}
	domain := []byte(in.Domain)

	msg := make([]byte, 0, 2+len(signDataSchemaPrefix)+4+32+4+len(domain)+8+3+4+len(payload))
	msg = append(msg, 0xff, 0xff)
	msg = append(msg, signDataSchemaPrefix...)
	msg = append(msg, encodeWorkchainA(in.Workchain)...)
	msg = append(msg, in.AddressHash[:]...)
	msg = append(msg, encodeDomainLenA(uint32(len(domain)))...)
	msg = append(msg, domain...)
	msg = append(msg, encodeTimestampA(in.Timestamp)...)
	msg = append(msg, tag...)
	msg = append(msg, encodePayloadLenA(uint32(len(payload)))...)
	msg = append(msg, payload...)
	return sha256.Sum256(msg)
}

// Contract A field encoders — big-endian. Deliberately NOT shared with ton_proof.go.
func encodeWorkchainA(wc int32) []byte {
	b := make([]byte, 4)
	binary.BigEndian.PutUint32(b, uint32(wc))
	return b
}
func encodeDomainLenA(n uint32) []byte {
	b := make([]byte, 4)
	binary.BigEndian.PutUint32(b, n)
	return b
}
func encodeTimestampA(ts uint64) []byte {
	b := make([]byte, 8)
	binary.BigEndian.PutUint64(b, ts)
	return b
}
func encodePayloadLenA(n uint32) []byte {
	b := make([]byte, 4)
	binary.BigEndian.PutUint32(b, n)
	return b
}

File: tc-v2-verify-rs/src/sign_data.rs
//! Contract A — `TC_V2_SIGNDATA_VERIFY_V1` (owner signature, signData).
//!
//! Big-endian length/timestamp, `txt`/`bin` type discriminator, SINGLE sha256.
//! This module owns its own field encoders (`encode_*` below). It shares NO
//! serialization/endian/hash-pipeline helper with `ton_proof.rs` — see the
//! prohibition in docs/spec/tc-v2-contract-boundaries.md. The conscious
//! duplication of `encode_domain_length` / `encode_timestamp` across the two
//! modules is the intended design, not debt.

use crate::sha256::sha256;
use crate::util::b64_decode;

const SCHEMA_PREFIX: &[u8] = b"ton-connect/sign-data/";

/// Owner-signature payload, per TC v2 `signData` type.
pub enum Payload<'a> {
    /// Signed as opaque UTF-8 bytes of the string.
    Text(&'a str),
    /// `bytes` field as standard base64; signed as the decoded bytes.
    Binary(&'a str),
}

pub struct SignDataInput<'a> {
    pub workchain: i32,
    pub address_hash: [u8; 32],
    pub domain: &'a str,
    pub timestamp: u64,
    pub payload: Payload<'a>,
}

/// Build the Contract A message and return its sha256 digest (the bytes ed25519 signs).
pub fn sign_data_digest(input: &SignDataInput) -> [u8; 32] {
    let (tag, payload): (&[u8], Vec<u8>) = match &input.payload {
        Payload::Text(t) => (b"txt", t.as_bytes().to_vec()),
        Payload::Binary(b64) => (b"bin", b64_decode(b64)),
    };
    let domain = input.domain.as_bytes();

    let mut m = Vec::new();
    m.extend_from_slice(&[0xff, 0xff]);
    m.extend_from_slice(SCHEMA_PREFIX);
    m.extend_from_slice(&encode_workchain(input.workchain));
    m.extend_from_slice(&input.address_hash);
    m.extend_from_slice(&encode_domain_length(domain.len() as u32));
    m.extend_from_slice(domain);
    m.extend_from_slice(&encode_timestamp(input.timestamp));
    m.extend_from_slice(tag);
    m.extend_from_slice(&encode_payload_length(payload.len() as u32));
    m.extend_from_slice(&payload);
    sha256(&m)
}

// --- Contract A field encoders (big-endian). NOT shared with Contract B. ---
fn encode_workchain(wc: i32) -> [u8; 4] {
    wc.to_be_bytes()
}
fn encode_domain_length(n: u32) -> [u8; 4] {
    n.to_be_bytes()
}
fn encode_timestamp(ts: u64) -> [u8; 8] {
    ts.to_be_bytes()
}
fn encode_payload_length(n: u32) -> [u8; 4] {
    n.to_be_bytes()
}
